resources/sdk/purecloudkotlin/templates/ApiClient.kt.mustache

```diff
@@ -409,6 +467,36 @@ class ApiClient private constructor(builder: Builder) : AutoCloseable {
         }
     }
 
+    @Throws(IOException::class, ApiException::class)
+    private fun handleExpiredAccessToken() {
+        // Attempt to refresh the access_token if a refresh isn't in progress and lock any other threads out of the method
+        if (refreshInProgress.compareAndSet(false, true)) {
+            try {
+                val refreshResponse = this.refreshCodeAuthorization()
+                if (refreshResponse.statusCode != 200) {
+                    val message = "error"
+                    throw ApiException(refreshResponse.statusCode, message, refreshResponse.headers as MutableMap<String, String>?, refreshResponse.rawBody)
+                }
+            } finally {
+                refreshInProgress.set(false)
+            }
+        } else {
+            var timeSpentWaiting = 0.0
+            val sleepDuration = 200
+            // Sleep for 200ms at a time for a maximum of refreshTokenWaitTime for other thread to complete refresh
+            while (timeSpentWaiting < refreshTokenWaitTime) {
+                try {
+                    Thread.sleep(sleepDuration.toLong())
+                } catch (e: InterruptedException) {
+                    Thread.currentThread().interrupt()
+                }
+                timeSpentWaiting += sleepDuration.toDouble() / 1000
```

Comment:
Adding sleeps together like this will miss some time and the total duration waited will be longer than calculated as it does not take into account processing time. Instead, set a variable with the current time before the loop and make the while condition something like `currentTime - startTime < refreshTokenWaitTime`. Using unix timestamps is one good option because using a single number to represent the timestamp makes calculating the duration very easy.